File: JavaScript/strategy-pattern/MiniDuckSimulator.js
import MallardDuck from './MallardDuck.js';
import ModelDuck from './ModelDuck.js';
import FlyRocketPowered from './FlyRocketPowered.js';

class MiniDuckSimulator {
  constructor() {
    this.mallard = undefined;
    this.model = undefined;
  }

  perform() {
    this.mallard = new MallardDuck();
    this.mallard.performFly();
    this.mallard.performQuack();

    this.model = new ModelDuck();
    this.model.performFly();
    this.model.performQuack();
    this.model.setFlyBehavior(new FlyRocketPowered());
    this.model.performFly();
  }
}

const simulator = new MiniDuckSimulator();
simulator.perform();

File: Ruby/strategy-pattern/mini_duck_simulator.rb
# frozen_string_literal: true

require_relative './mallard_duck'
require_relative './model_duck'
require_relative './fly_rocket_powered'

# MiniDuckSimulator Class
class MiniDuckSimulator
  def perform
    @mallard = MallardDuck.new
    @mallard.perform_fly
    @mallard.perform_quack

    @model = ModelDuck.new
    @model.perform_fly
    @model.perform_quack
    @model.fly_behavior = FlyRocketPowered.new
    @model.perform_fly
  end
end

simulator = MiniDuckSimulator.new
simulator.perform
